driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -201,19 +201,32 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> ou
             // version first bit is the "direction" of the frame (request or response)
             version = version & 0x7F;
 
-            DecoderForStreamIdSize decoder = (version >= 3) ? decoderV3 : decoderV1;
+            DecoderForStreamIdSize decoder = decoderForStreamIdSize.get();
+            if (decoder == null) {
```

Comment:
> you are right that Decoder should instantiate only one DecoderForStreamIdSize and use it consistently, so I'm ok with your approach as well.

Sounds good, I think if we can count on C* always sending the same protocol version on a connection this approach works, but added an assertion just to see if that would ever be a problem (it helped me capture the bug i had previously where I passed in the protocolVersion originally).   Wasn't sure if using an `AtomicReference` was required or not.  I'd expect decode to only ever be called one at a time, but wasn't sure if that was a guarantee netty makes.